Framework/LiveData/src/Kafka/KafkaEventStreamDecoder.cpp

```diff
@@ -219,37 +300,154 @@ void KafkaEventStreamDecoder::captureImplExcept() {
+  /* Store the buffered events */
+  std::transform(
+      detData.begin(), detData.end(), detData.begin(),
+      std::back_inserter(m_receivedEventBuffer),
+      [&](uint64_t detId, uint64_t tof) -> BufferedEvent {
+        return {m_specToIdx[static_cast<int>(detId)], tof, pulseIndex};
+      });
+
+  const auto endTime = std::chrono::system_clock::now();
+  const std::chrono::duration<double> dur = endTime - starttime;
+  totalEventFromMessageDuration += dur.count();
+  numEventFromMessageCalls += 1;
+}
+
+void KafkaEventStreamDecoder::populateEventDataWorkspace() {
+  g_log.notice() << "Populating event workspace with "
+                 << m_receivedEventBuffer.size() << " events\n";
+
+  const auto startTime = std::chrono::system_clock::now();
+
+  /* Parallel sort received event buffer */
+  tbb::parallel_sort(
+      m_receivedEventBuffer.begin(), m_receivedEventBuffer.end(),
+      [&](const BufferedEvent &lhs, const BufferedEvent &rhs) -> bool {
+        const auto &lhsPulse = m_receivedPulseBuffer[lhs.pulseIndex];
+        const auto &rhsPulse = m_receivedPulseBuffer[rhs.pulseIndex];
+
+        /* If events are from different periods compare the period
+         * numbers, otherwise compare the workspace index */
+        return (lhsPulse.periodNumber != rhsPulse.periodNumber)
+                   ? lhsPulse.periodNumber < rhsPulse.periodNumber
+                   : lhs.wsIdx < rhs.wsIdx;
+      });
+
+  /* Compute groups for parallel insertion */
+  const auto numberOfGroups = omp_get_max_threads();
+  std::vector<size_t> groupBoundaries(numberOfGroups + 1);
+
+  /* First group always starts at beginning of buffer */
+  groupBoundaries[0] = 0;
+
+  /* Iterate over groups */
+  for (auto group = 1; group < numberOfGroups; ++group) {
+    /* Calculate a reasonable end boundary for the group */
+    groupBoundaries[group] =
+        m_receivedEventBuffer.size() / numberOfGroups * group;
+
+    /* Advance the end boundary of the group until all events for a given
+     * workspace index fall within a single group */
+    while (groupBoundaries[group] + 1 < m_receivedEventBuffer.size() &&
+           (m_receivedEventBuffer[groupBoundaries[group]].wsIdx ==
+            m_receivedEventBuffer[groupBoundaries[group] + 1].wsIdx)) {
+      ++groupBoundaries[group];
+    }
+
+    /* TODO: what was this for? */
+    ++groupBoundaries[group];
+  }
+
+  /* Last group always finishes at end of buffer */
+  groupBoundaries[numberOfGroups] = m_receivedEventBuffer.size();
+
+/* Insert events into EventWorkspace(s) */
+#pragma omp parallel for shared(groupBoundaries, m_localEvents)
+  for (auto group = 0; group < numberOfGroups; ++group) {
+    for (auto idx = groupBoundaries[group]; idx < groupBoundaries[group + 1];
+         ++idx) {
+      const auto &event = m_receivedEventBuffer[idx];
+      const auto &pulse = m_receivedPulseBuffer[event.pulseIndex];
+      auto &spectrum =
+          m_localEvents[pulse.periodNumber]->getSpectrum(event.wsIdx);
```

Comment:
`getSpectrum` is quite expensive, can you move this out of the inner loop safely?